src/crypto/tls/delegated_credentials_test.go

```diff
@@ -0,0 +1,413 @@
+	}
+}
+
+// Test encoding/decoding of delegated credentials.
+func TestDelegatedCredentialMarshal(t *testing.T) {
+	initialize()
+	cert := dcTestCerts["dc"]
+	time := dcNow.Sub(cert.Leaf.NotBefore) + dcMaxTTL
+
+	delegatedCred, _, err := NewDelegatedCredential(cert, ECDSAWithP256AndSHA256, time, "server")
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	ser, err := delegatedCred.Marshal()
+	if err != nil {
+		t.Error(err)
+	}
+
+	delegatedCred2, err := UnmarshalDelegatedCredential(ser)
+	if err != nil {
+		t.Error(err)
+	}
+
+	err = equal(delegatedCred, delegatedCred2)
+	if err != nil {
+		t.Error(err)
+	}
+
+	if delegatedCred.Algorithm != delegatedCred2.Algorithm {
+		t.Errorf("scheme mismatch: got %04x; want %04x", delegatedCred2.Algorithm, delegatedCred.Algorithm)
+	}
+
+	if !bytes.Equal(delegatedCred2.Signature, delegatedCred.Signature) {
+		t.Error("Signature mismatch")
+	}
+}
+
+var dcTests = []struct {
+	clientDCSupport  bool
+	serverDCSupport  bool
+	clientSkipVerify bool
+	clientMaxVers    uint16
+	serverMaxVers    uint16
+	expectSuccess    bool
+	expectDC         bool
+	name             string
+}{
+	{true, true, false, VersionTLS13, VersionTLS13, true, true, "tls13"},
+	{true, true, true, VersionTLS13, VersionTLS13, true, true, "tls13, using server skip verify"},
+	{false, true, false, VersionTLS13, VersionTLS13, true, false, "client no DC"},
+	{true, false, false, VersionTLS13, VersionTLS13, true, false, "server no DC"},
+	{true, false, false, VersionTLS12, VersionTLS13, true, false, "client using TLS 1.2. No DC is supported in that version."},
+	{true, false, false, VersionTLS13, VersionTLS12, true, false, "server using TLS 1.2. No DC is supported in that version."},
+	{true, false, false, VersionTLS11, VersionTLS13, true, false, "client using TLS 1.1. No DC is supported in that version."},
+	{true, false, false, VersionTLS13, VersionTLS10, false, false, "server using TLS 1.0. No DC is supported in that version."},
+}
+
+// Checks that the client suppports a version >= 1.3 and accepts delegated
+// credentials. If so, it returns the delegation certificate; otherwise it
+// returns a non-delegated certificate.
+func testServerGetCertificate(ch *ClientHelloInfo) (*Certificate, error) {
+	versOk := false
+	for _, vers := range ch.SupportedVersions {
+		versOk = versOk || (vers >= uint16(VersionTLS13))
+	}
+
+	ch.SupportDelegatedCredential = true
+	if versOk && ch.SupportDelegatedCredential {
+		return dcTestCerts["dc"], nil
+	}
+	return dcTestCerts["no dc"], nil
+
+}
+
+// Checks that the client supports the signature algorithm supported by the test
+// server, and that the server has a Delegated Credential.
+func testServerGetDelegatedCredential(ch *ClientHelloInfo) (*DelegatedCredential, crypto.PrivateKey, error) {
+	schemeOk := false
+	for _, scheme := range ch.SignatureSchemes {
+		schemeOk = schemeOk || (scheme == dcTestDCScheme)
+	}
+
+	ch.SupportDelegatedCredential = true
```

Comment:
As Peter noted, why does need to modify `ch`?

Reply:
I'll investigate this later.

Reply:
Check commit 7b0e3dc, which correctly inputs how it gets populated @chris-wood